RevDiff: Show common files for BASE->A

gerhardol opened the pull request:
Fixes #4154
WIP, based on #7825 
Some cleanups are planned as a follow up to this PR

Thanks to @mdonatas for the issue and an initial implementation. I have not used any of that code, but view the initial implementation.

## Proposed changes

See #4154 for a description, this PR is to improve comparing changes between branches.
#7720 adds the diff from the common BASE to the selected revision. This also shows what files that are unique to the first/selected commits, as well the files that were changed in the same way.

FileStatusList has until now basically handled the files, selected revision and the (possibly multiple) first or parents as separate data, occasionally bundled/unbundled.

To handle the possibly multiple a->b diff pairs, the datastructures are changed to internally use datastructure including the selected revision in GitItemStatusWithParent to allow multiple selected revisions and simplify handling in other forms

To simplify the code review, this has been split in several commits. The last three could be separated to a separate PR.
 * Refactoring to use ObjectId instead of strings in one more situation.
 * The changes to show unique and common diff from BASE to the selected(b) commit mostly requires changes in FileStatusList.
 * Refactor: Prepare to support multiple selected versions. To also show BASE to first, other forms need to be aware of the changes too (really just RevDiff).
 * RevDiff: Show Unique files BASE->first(A). This is separated (one line), as I am not convinced that this information is understood.
 * Refactor: SelectedItems -> SelectedItemsWithParent. Where possible, use the structures with selected/parent revision, instead of implicit information in the FileStatusList. This will be an enabler for other changes.

Some improvement areas (discussed in the issue):
 * Show groups also if no files
 * tooltips to show the contents better
 * Collapse other than first->selected by default
 * (async loading, show results earlier)

## Screenshots

### Before

![image](https://user-images.githubusercontent.com/6248932/77567063-e69f4900-6ec6-11ea-8357-d75ffeac9782.png)

### After

The BASE -> first unique files diff is marked below

![image](https://user-images.githubusercontent.com/6248932/77566891-9d4ef980-6ec6-11ea-9e97-319a40284a90.png)

## Test methodology 
Manual


----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
CodeFactor complains at ExecuteCommand(), it natural to have many return points there so no action planned.

gerhardol commented:
Will be rebased on #7908 when that is merged
Note: I could factor out the show common/unique diff parts separately to a separate (quite small) PR if desired and leave the refactoring in this PR and add unique BASE->First to a separate PR (one line) if that simplifies reviews.

gerhardol commented:
Rebased on #7923 - reviewable (merge probable wait to after 4.0)

gerhardol commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -41,8 +42,8 @@ public sealed partial class FileStatusList : GitModuleControl
         private Rectangle _dragBoxFromMouseDown;
-        private IReadOnlyList<(GitRevision revision, IReadOnlyList<GitItemStatus> statuses)> _itemsWithParent = Array.Empty<(GitRevision, IReadOnlyList<GitItemStatus>)>();
-        private IReadOnlyList<(GitRevision revision, string description, IReadOnlyList<GitItemStatus> statuses)> _itemsWithDescription = Array.Empty<(GitRevision, string, IReadOnlyList<GitItemStatus>)>();
+        private IReadOnlyList<(GitRevision parentRev, GitRevision selectedRev, string summary, IReadOnlyList<GitItemStatus> statuses)> _itemsWithDescription =
+            Array.Empty<(GitRevision, GitRevision, string, IReadOnlyList<GitItemStatus>)>();
```
Applying to #7923

gerhardol commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -195,8 +196,8 @@ private static SortDiffListContextMenuItem CreateSortByContextMenuItem()
         [Browsable(false)]
-        public IEnumerable<GitItemStatus> AllItems =>
-            FileStatusListView.ItemTags<GitItemStatus>();
+        public IEnumerable<GitItemStatusWithParent> AllItems =>
+            FileStatusListView.ItemTags<GitItemStatusWithParent>();
```
Applying to #7923 - one line

gerhardol commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -266,35 +267,25 @@ private bool FilterVisibleInternal
         [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
         [Browsable(false)]
         public IReadOnlyList<GitItemStatus> GitItemFilteredStatuses =>
-            AllItems.AsReadOnlyList();
+            AllItems.Select(i => i.Item).AsReadOnlyList();
```
Applying to #7923 - one line

gerhardol commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -382,25 +366,32 @@ ListViewItem GetItemByStatus(GitItemStatus status)
-                .Select(i => i.Group?.Tag<GitRevision>())
-                .Where(r => r != null)
-                .Distinct(); // Parents should be distinct to avoid iterating same items
+        public GitItemStatusWithParent SelectedItemWithParent => FileStatusListView.LastSelectedItem()?.Tag<GitItemStatusWithParent>();
```
ok

gerhardol commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -658,91 +642,120 @@ public void SelectStoredNextIndex(int defaultIndex = -1)
 
-            ObjectId GetRevisionOrHead(ObjectId rev, Lazy<ObjectId> head)
+            static ObjectId GetRevisionOrHead(GitRevision rev, Lazy<ObjectId> head)
             {
-                return rev.IsArtificial ? head.Value : rev;
+                return rev.IsArtificial ? head.Value : rev.ObjectId;
             }
```
Applying to #7923

gerhardol commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -1724,7 +1734,7 @@ public GitStatusListSorter(IComparer<GitItemStatus> gitStatusItemSorter = null)
 
             public override int Compare(ListViewItem x, ListViewItem y)
             {
-                return StatusComparer.Compare(x.Tag as GitItemStatus, y.Tag as GitItemStatus);
+                return StatusComparer.Compare((x.Tag as GitItemStatusWithParent).Item, (y.Tag as GitItemStatusWithParent).Item);
             }
```
Applying to #7923

gerhardol commented on GitUI/UserControls/GitItemStatusWithParent.cs:
```diff
@@ -4,13 +4,23 @@ namespace GitUI.UserControls
+        /// First (Parent or A in diff)
+        /// </summary>
         public GitRevision ParentRevision { get; }
         public GitItemStatus Item { get; }
+        public override string ToString() => Item.ToString();
```
ok

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -784,24 +799,23 @@ private bool DeleteSelectedFiles()
+                var selected = DiffFiles.SelectedItemWithParent;
+                if (selected?.Item == null || selected.SelectedRevision == null || !selected.SelectedRevision.IsArtificial ||
                     MessageBox.Show(this, _deleteSelectedFiles.Text, _deleteSelectedFilesCaption.Text, MessageBoxButtons.YesNo,
-                                    MessageBoxIcon.Warning) !=
+                        MessageBoxIcon.Warning) !=
                     DialogResult.Yes)
```
ok

mstv commented:
1. The issue with FormStash is still present. This seems to be new with this PR.

2. I could not build the branch `gerhardol/feature/i7338-revdiff-stage-2` https://github.com/gitextensions/gitextensions/commit/dc32161737cb4065e2f21bc79d50bbcf74208311, which is based on this PR. (IIUIC, you mentioned you are waiting for this PR to be reviewed before fixing this.)

---

3. Sometimes the diffs with BASE are missing (both with this PR and already with current master):

![grafik](https://user-images.githubusercontent.com/36601201/79507767-91bfa000-8038-11ea-88e2-2746f81f0b90.png)

It is not caused by a direct common parent:

![grafik](https://user-images.githubusercontent.com/36601201/79508124-34781e80-8039-11ea-9ca5-c1359bb4dc26.png)

---

4. Empty merges (to master) can be hard to recognize (seen at work :unamused:).
Will be easier if all groups are displayed. Should we extend the `ExListView` to show dummy entries?
Or should we add a dummy `GitItemStatus` like in https://github.com/gitextensions/gitextensions/commit/de81cc78c26a24cf9181fe8bfcb16078573dccda?  

![grafik](https://user-images.githubusercontent.com/36601201/79510715-25479f80-803e-11ea-9fbf-544a2fce45c7.png)

mstv commented:
NRE is thrown sometimes when opening the context menu quickly for different files in FileStatusList
maybe with `Esc` hit inbetween

## Error Details
```
System.NullReferenceException: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   bei GitUI.FileStatusList.FileStatusListView_ContextMenu_Opening(Object sender, CancelEventArgs e)
   bei System.ComponentModel.CancelEventHandler.Invoke(Object sender, CancelEventArgs e)
   bei System.Windows.Forms.ToolStripDropDown.OnOpening(CancelEventArgs e)
   bei System.Windows.Forms.ToolStripDropDown.SetVisibleCore(Boolean visible)
   bei System.Windows.Forms.ContextMenuStrip.SetVisibleCore(Boolean visible)
   bei System.Windows.Forms.ToolStripDropDown.Show(Control control, Point position)
   bei System.Windows.Forms.ContextMenuStrip.ShowInternal(Control source, Point location, Boolean isKeyboardActivated)
   bei System.Windows.Forms.Control.WmContextMenu(Message& m, Control sourceControl)
   bei System.Windows.Forms.Control.WmContextMenu(Message& m)
   bei System.Windows.Forms.Control.WndProc(Message& m)
   bei System.Windows.Forms.ListView.WndProc(Message& m)
   bei GitUI.UserControls.ExListView.WndProc(Message& m)
   bei System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
   bei System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   bei System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```

## Environment
- Git Extensions 3.3.0
- Build 147374d0960b2e634d4b84a5239130f39c954709 (Dirty)
- Git 2.21.0.windows.1 (recommended: 2.25.1 or later)
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.8.4069.0
- DPI 96dpi (no scaling)

gerhardol commented:
I wanted to have #6825 merged first, but reviews took some time and the @RussKie did not like the API.

Since I was working on this and related issues meanwhile there has been many conflicts, I have spent much too much time breaking out smaller parts and rebasing.
To be clear, I am not complaining that changes are not accepted (RussKie is doing a great job with most things), but obviously I should not have put time on this and should have put time on another project.

If the PRs are bundled, there will be much to many changes and they cannot be reviewed due to that.

> Blocking the merge until after the release.

Of course

>    The issue with FormStash is still present. This seems to be new with this PR.

Fixing, the simplifications in the cleanup needs a null check still. I will remove the waitcursor here too.

>    I could not build the branch gerhardol/feature/i7338-revdiff-stage-2 dc32161, which is based on this PR. (IIUIC, you mentioned you are waiting for this PR to be reviewed before fixing this.)

While I appreciate your and other reviews, it is @RussKie that will have requests that breaks all related PRs so I am waiting on that.
Have updated feature/i7338-revdiff-stage-2, my private eat-your-own-dogfood only have a few patches on top of that so maybe I can push feature/i7338-revdiff-stage-2 too. As it seems now, this is how merge will be done anyway.

>    Sometimes the diffs with BASE are missing (both with this PR and already with current master):

For these commits, I see BASE. Any more info to recreate?

>    Empty merges (to master) can be hard to recognize (seen at work 😒).
    Will be easier if all groups are displayed. Should we extend the ExListView to show dummy entries?
    Or should we add a dummy GitItemStatus like in de81cc7?

This applies to all groups. I would like to show also empty groups and collapse other than the first.
Another improvement with collapsing would be to run the diffs async, to improve response time.
(Info could be added to tooltips)

We can have the discussion here, but I see this as something separate.

> NRE is thrown sometimes when opening the context menu quickly for different files in FileStatusList
maybe with Esc hit inbetween

Similar to the stash issue, fixing

RussKie commented:
Are you open for a voice chat about this?

I've been flat out busy with work, and I have neglected your work here,
sorry, which obviously led to this frustrating situation.



On Sat, Apr 18, 2020, 11:06 PM Gerhard Olsson <notifications@github.com>
wrote:

> I wanted to have #6825
> <https://github.com/gitextensions/gitextensions/issues/6825> merged
> first, but reviews took some time and the @RussKie
> <https://github.com/RussKie> did not like the API.
>
> Since I was working on this and related issues meanwhile there has been
> many conflicts, I have spent much too much time breaking out smaller parts
> and rebasing.
> To be clear, I am not complaining that changes are not accepted (RussKie
> is doing a great job with most things), but obviously I should not have put
> time on this and should have put time on another project.
>
> If the PRs are bundled, there will be much to many changes and they cannot
> be reviewed due to that.
>
> Blocking the merge until after the release.
>
> Of course
>
> The issue with FormStash is still present. This seems to be new with this
> PR.
>
> Fixing, the simplifications in the cleanup needs a null check still. I
> will remove the waitcursor here too.
>
> I could not build the branch gerhardol/feature/i7338-revdiff-stage-2
> dc32161
> <https://github.com/gitextensions/gitextensions/commit/dc32161737cb4065e2f21bc79d50bbcf74208311>,
> which is based on this PR. (IIUIC, you mentioned you are waiting for this
> PR to be reviewed before fixing this.)
>
> While I appreciate your and other reviews, it is @RussKie
> <https://github.com/RussKie> that will have requests that breaks all
> related PRs so I am waiting on that.
> Have updated feature/i7338-revdiff-stage-2, my private
> eat-your-own-dogfood only have a few patches on top of that so maybe I can
> push feature/i7338-revdiff-stage-2 too. As it seems now, this is how merge
> will be done anyway.
>
> Sometimes the diffs with BASE are missing (both with this PR and already
> with current master):
>
> For these commits, I see BASE. Any more info to recreate?
>
> Empty merges (to master) can be hard to recognize (seen at work 😒).
> Will be easier if all groups are displayed. Should we extend the
> ExListView to show dummy entries?
> Or should we add a dummy GitItemStatus like in de81cc7
> <https://github.com/gitextensions/gitextensions/commit/de81cc78c26a24cf9181fe8bfcb16078573dccda>
> ?
>
> This applies to all groups. I would like to show also empty groups and
> collapse other than the first.
> Another improvement with collapsing would be to run the diffs async, to
> improve response time.
> (Info could be added to tooltips)
>
> We can have the discussion here, but I see this as something separate.
>
> NRE is thrown sometimes when opening the context menu quickly for
> different files in FileStatusList
> maybe with Esc hit inbetween
>
> Similar to the stash issue, fixing
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7899#issuecomment-615867161>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXXBKFD34OZQDMYV64DRNGQUTANCNFSM4LTVGY4Q>
> .
>

mstv commented:
Thank you for the updated branches, Gerhard.

> Sometimes the diffs with BASE are missing (both with this PR and already with current master):

This is the case for diffs of an artifical commit with a branch based on HEAD.

![grafik](https://user-images.githubusercontent.com/36601201/79507767-91bfa000-8038-11ea-88e2-2746f81f0b90.png)

> This applies to all groups. I would like to show also empty groups and collapse other than the first.
Another improvement with collapsing would be to run the diffs async, to improve response time.
(Info could be added to tooltips)

👍

> We can have the discussion here, but I see this as something separate.

Yes, perhaps even https://github.com/gitextensions/gitextensions/issues/1326#issuecomment-602500732

mstv commented on GitUI/UserControls/GitItemStatusWithParent.cs:
```diff
@@ -4,13 +4,23 @@ namespace GitUI.UserControls
     public sealed class GitItemStatusWithParent
     {
-        public GitItemStatusWithParent(GitRevision parent, GitItemStatus item)
+        public GitItemStatusWithParent(GitRevision parent, GitRevision selected, GitItemStatus item)
```
> GitItemStatusWithParent should maybe be renamed. mdonatas used FileStatusItem, also 'collapsed' access to Item (e.g. Name => GitItemStatus.Name).
Also SelectedRevision -> SecondRevision and ParentRevision - > FirstRevision

👍

gerhardol commented:
Discussed with @RussKie , my summary:
I will rename according to https://github.com/gitextensions/gitextensions/pull/7899#pullrequestreview-398182585, then squash and make a final check before merge before next release. #7825 can then be formally rebased on this and the API discussion there be continued.

The exceptions that @mstv reported is not a big concern, the handling is unified in this and related PRs and is reviewable, but if doing the update many times just to keep up, I may not do the review so carefully as when close to merge.

gerhardol commented:
Refactored the naming according to the suggestion including the internal parameters and made a few more conversions to FileStatusItem. No functional change, but a few cleanups
(I consider eliminating the getters for SelectedGitItem/SelectedGitItems too, to cleanup the interface for FileStatusList.)

@mstv, if you agree, I would like to merge this before 4.0. I would like to not have to rebase this branch and resolve conflicts again...
Commits will be squashed, except ```RevDiff: Show Unique files BASE->first(A)```

I also updated feature/i7338-revdiff-stage-2, will move feature/i7338-revdiff-stage to this when merged

RussKie commented on GitUI/UserControls/FileStatusItem.cs:
```diff
@@ -0,0 +1,26 @@
+        /// </summary>
+        public GitRevision FirstRevision { get; }
+        public GitItemStatus Item { get; }
+        public override string ToString() => Item.ToString();
```
❗ Possible NRE.
We either need to validate and throw in the ctor, or be defensive here.

Tests? :D

gerhardol replied:
See below

RussKie commented on GitUI/UserControls/FileStatusItem.cs:
```diff
@@ -0,0 +1,26 @@
+{
+    public sealed class FileStatusItem
+    {
+        public FileStatusItem(GitRevision firstRev, GitRevision secondRev, GitItemStatus item)
+        {
+            SecondRevision = secondRev;
+            FirstRevision = firstRev;
+            Item = item;
+        }
```
❗  These all can be `null`, and must either must validate and throw, if not allowed, or be defensive below and handle `null`s at call sites

gerhardol replied:
Defensive, similar to before when the information (item/revs) was retrieved from several sources. So null must be checked where used, something I have looked for (but missed the internal usage above).
But this could be improved on. Item should be NotNull, need to review the handling first though.
Revisions are set to null in a few situations currently. This can likely be changed for second (selected) revision, but not for first. Generally first is of no interest for New items and irelevant for the first commit. A random commit hash could be used, but then I prefer null.

gerhardol replied:
Updating and added annotations

gerhardol replied:
Changed this for SelectedRevision too

RussKie commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -1467,9 +1435,9 @@ private void FileStatusListView_MouseMove(object sender, MouseEventArgs e)
                     var fileList = new StringCollection();
 
-                    foreach (GitItemStatus item in SelectedItems)
+                    foreach (var item in SelectedItems)
```
This change is undesirable as it reduces comprehensibility of the code.

gerhardol replied:
I thought the recommendation was to use ```var``` where type was defined, but I can change (to FileStatusItem in these examples). Not sure if there are more though.

RussKie commented on GitUI/CommandsDialogs/FormCommit.cs:
```diff
@@ -1625,38 +1624,25 @@ private void UnstagedSelectionChanged(object sender, EventArgs e)
-            _currentSelection = Unstaged.SelectedItems.ToList();
-            GitItemStatus item = Unstaged.SelectedItem;
+            _currentSelection = Unstaged.SelectedGitItems.ToList();
+            var item = Unstaged.SelectedItem;
```
ditto

mstv commented:
Right now, I have noticed a NRE for an empty FileStatusList on `F3` and on opening the context menu.

![grafik](https://user-images.githubusercontent.com/36601201/80319801-d0acdd00-8812-11ea-9551-0222d183abaf.png)

## Error Details
```
System.NullReferenceException: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   bei GitUI.CommandsDialogs.RevisionDiffController.ShouldShowMenuDeleteFile(ContextMenuSelectionInfo selectionInfo)
   bei GitUI.CommandsDialogs.RevisionDiffControl.UpdateStatusOfMenuItems()
   bei GitUI.CommandsDialogs.RevisionDiffControl.ExecuteCommand(Int32 cmd)
   bei GitUI.CommandsDialogs.RevisionDiffControl.ExecuteCommand(Command cmd)
   bei GitUI.CommandsDialogs.FormBrowse.<ExecuteCommand>g__OpenWithDifftool|163_5()
   bei GitUI.CommandsDialogs.FormBrowse.ExecuteCommand(Int32 cmd)
   bei ResourceManager.GitExtensionsFormBase.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   bei ResourceManager.GitExtensionsControl.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   bei System.Windows.Forms.Control.PreProcessMessage(Message& msg)
   bei System.Windows.Forms.Control.PreProcessControlMessageInternal(Control target, Message& msg)
   bei System.Windows.Forms.Application.ThreadContext.PreTranslateMessage(MSG& msg)
```

## Environment
- Git Extensions 4.0.0.9295
- Build 98c9d5dcc3e2235fb3ea064673d73c9626848ab5 (Dirty)
- Git 2.26.0.windows.1
- Microsoft Windows NT 6.2.9200.0
- .NET Framework 4.8.4121.0
- DPI 96dpi (no scaling)

gerhardol commented:
> Right now, I have noticed a NRE for an empty FileStatusList on `F3` and on opening the context menu.

Consequence of rebasing on #8012
I had noticed myself but thought it was in master too and was to submit a PR 

Added a test, I considered changing the defaults to no selection (now it is some selection), but too many changes for this fix

(Another reason to why I do not want to continue rebasing this PR...)

RussKie commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -395,7 +395,7 @@ private void DiffFiles_SelectedIndexChanged(object sender, EventArgs e)
         {
             FileStatusItem item = DiffFiles.SelectedItem;
-            if (item?.SecondRevision == null || item.Item == null || !item.Item.IsTracked)
+            if (item?.SecondRevision == null || !item.Item.IsTracked)
```
We are 100% certain `item.Item` can never be `null` at this point, right?

gerhardol replied:
If it is null, many other actions will be broken
This is when invoking actions on the selections so also item should not be null (but there has been issues...)

The risk with the change is when creating the FileStatusItem where GitStatusItem is null but not used anyway. This is why SecondRevision is [CanBeNull], ViewPullRequestsForm has special handling that never sets or accesses the first/second revision and I prefer to not add a dummy revision.

mstv commented on GitUI/CommandsDialogs/RepoHosting/ViewPullRequestsForm.cs:
```diff
@@ -381,9 +391,8 @@ private void SplitAndLoadDiff(string diffData)
-            // patches directly from the cache
-            _fileStatusList.SetDiffs(selectedRev: null, parentRev: null, items: giss);
+            // Note: Commits in PR may not exist in the local repo
+            _fileStatusList.SetDiffs(firstRev: firstRev, secondRev: secondRev, items: giss);
```
nit: If arguments are named identically, I would omit the arguement names.
```suggestion
            _fileStatusList.SetDiffs(firstRev, secondRev, items: giss);
```

gerhardol replied:
My motivation for keeping the names is that these have changed, but let us assume this is stable now...

gerhardol commented on GitUI/CommandsDialogs/FormCommit.cs:
```diff
@@ -1625,38 +1624,25 @@ private void UnstagedSelectionChanged(object sender, EventArgs e)
-            {
-                Unstaged.ContextMenuStrip = UnstagedSubmoduleContext;
-            }
+            Unstaged.ContextMenuStrip = (item?.Item?.IsSubmodule ?? false) ? UnstagedSubmoduleContext : UnstagedFileContext;
```
ok

gerhardol commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -1304,7 +1271,8 @@ private void FileStatusListView_ContextMenu_Opening(object sender, CancelEventAr
             }
 
-            bool isSubmoduleSelected = SelectedItem != null && SelectedItem.IsSubmodule;
+            bool isSubmoduleSelected = SelectedItem?.Item?.IsSubmodule ?? false;
```
ok

gerhardol commented:
Rebased and squashed including the review comments, but also a minor simplification in FileStatusList:SetDiffs()
Preparing merge - if the fixup! merge is removed this will have to be rebased again

gerhardol commented:
@msftbot merge in 48 hours
